models/lfs.go

```diff
@@ -162,23 +169,39 @@ func LFSAutoAssociate(metas []*LFSMetaObject, user *user_model.User, repoID int6
 		oidMap[meta.Oid] = meta
 	}
 
-	cond := builder.NewCond()
 	if !user.IsAdmin {
-		cond = builder.In("`lfs_meta_object`.repository_id",
-			builder.Select("`repository`.id").From("repository").Where(accessibleRepositoryCondition(user)))
-	}
-	newMetas := make([]*LFSMetaObject, 0, len(metas))
-	if err := sess.Cols("oid").Where(cond).In("oid", oids...).GroupBy("oid").Find(&newMetas); err != nil {
-		return err
-	}
-	for i := range newMetas {
-		newMetas[i].Size = oidMap[newMetas[i].Oid].Size
-		newMetas[i].RepositoryID = repoID
-	}
-	if err := db.Insert(ctx, newMetas); err != nil {
-		return err
+		newMetas := make([]*LFSMetaObject, 0, len(metas))
+		cond := builder.In(
+			"`lfs_meta_object`.repository_id",
+			builder.Select("`repository`.id").From("repository").Where(accessibleRepositoryCondition(user)),
+		)
+		err = sess.Cols("oid").Where(cond).In("oid", oids...).GroupBy("oid").Find(&newMetas)
+		if err != nil {
+			return err
+		}
+		if len(newMetas) != len(oidMap) {
+			return fmt.Errorf("unable collect all LFS objects from database, expected %d, actually %d", len(oidMap), len(newMetas))
+		}
+		for i := range newMetas {
+			newMetas[i].Size = oidMap[newMetas[i].Oid].Size
+			newMetas[i].RepositoryID = repoID
+		}
+		if err = db.Insert(ctx, newMetas); err != nil {
+			return err
+		}
+	} else {
+		// admin can associate any LFS object to any repository, and we do not care about errors (eg: duplicated unique key),
+		// even if error occurs, it won't hurt users and won't make things worse
+		for i := range metas {
+			_, err = sess.Insert(&LFSMetaObject{
+				Pointer:      lfs.Pointer{Oid: metas[i].Oid, Size: metas[i].Size},
+				RepositoryID: repoID,
+			})
+			if err != nil {
+				log.Warn("failed to insert LFS meta object into database, err=%v", err)
+			}
```

Comment:
If the content store does not contain the LFS meta ID won't this result in dangling pointers and corruption of the content store? Shouldn't we be checking that the physical store contains something with this ID?

Reply:
The `metas` comes from the `pointerResult` slice in `routers/web/repo/lfs.go`, it has been check by `result.Exists, err = contentStore.Exists(pointerBlob.Pointer)`.

The old logic is: get a valid pointerResult slice -> generate a form with `oid size` list -> submit form -> parse the `oid size` list to `metas` -> insert into database. I didn't change these code (althought that's quite hacky).